pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+		delayBlocks:              5,
+		delayBlocksBumpFrequency: 100,
+		delayBlocksBumpFactor:    20,
+	}
+}
+
+// dkgAttemptParams represents parameters of a DKG attempt.
+type dkgAttemptParams struct {
+	index           uint
+	startBlock      uint64
+	excludedMembers []group.MemberIndex
+}
+
+// dkgAttemptFn represents a function performing a DKG attempt.
+type dkgAttemptFn func(*dkgAttemptParams) (*dkg.Result, error)
+
+// start begins the DKG retry loop using the given DKG attempt function.
+// The retry loop terminates when the DKG result is produced or the ctx
+// parameter is done, whatever comes first.
+func (drl *dkgRetryLoop) start(
+	ctx context.Context,
+	dkgAttemptFn dkgAttemptFn,
+) (*dkg.Result, error) {
+	// All selected operators should be qualified for the first attempt.
+	qualifiedOperatorsSet := drl.selectedOperators.Set()
+
+	for {
+		drl.attemptCounter++
+
+		// Check the loop stop signal.
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf(
+				"dkg retry loop received stop signal on attempt [%v]",
+				drl.attemptCounter,
+			)
+		}
+
+		// In order to start attempts >1 in the right place, we need to
+		// determine how many blocks were taken by previous attempts. We assume
+		// the worst case that each attempt failed at the end of the DKG
+		// protocol. That said, we need to increment the previous attempt start
+		// block by the number of blocks equal to the protocol duration and
+		// by some additional delay blocks. We need a small fixed delay in
+		// order to mitigate all corner cases where the actual attempt duration
+		// was slightly longer than the expected duration determined by the
+		// dkg.ProtocolBlocks function. For example, the attempt may fail at
+		// the end of the protocol but the error is returned after some time
+		// and more blocks than expected are mined in the meantime.
+		// Additionally, we want to strongly extend the delay period
+		// periodically in order to give some additional time for nodes to
+		// recover and re-fill their internal TSS pre-parameters pools.
+		if drl.attemptCounter > 1 {
+			delayBlocks := drl.delayBlocks
+			if drl.attemptCounter%drl.delayBlocksBumpFrequency == 0 {
+				delayBlocks *= drl.delayBlocksBumpFactor
+			}
+
+			drl.attemptStartBlock = drl.attemptStartBlock +
+				dkg.ProtocolBlocks() +
+				delayBlocks
+		}
+
+		// Exclude all members controlled by the operators that were not
+		// qualified for the current attempt.
+		excludedMembers := make([]group.MemberIndex, 0)
+		attemptSkipped := false
+		for i, operator := range drl.selectedOperators {
+			if !qualifiedOperatorsSet[operator] {
+				memberIndex := group.MemberIndex(i + 1)
+				excludedMembers = append(excludedMembers, memberIndex)
+
+				// If the given member was not qualified for the given attempt,
+				// mark this attempt as skipped in order to skip the execution
+				// and set up the next attempt properly.
+				if memberIndex == drl.memberIndex {
+					attemptSkipped = true
```

Comment:
I think that if we hit this condition, we don't need to continue populating `excludedMembers`. Is it possible to break a `for range`?

Reply:
Yep we can break here: 054c6cda